Fix dispatch issues

tomjnixon opened the pull request:
See #632, #633 and #634.

This doesn't cover the documentation or cmake aspects of #633; i think that's a separate job.

I'm not totally sure about the consequences of "add explicit check for architecture support in batch"; it could make for more mysterious error messages if simd_register is used directly for some reason, but I think it'll be alright.

serge-sans-paille commented on include/xsimd/types/xsimd_register.hpp:
```diff
@@ -20,7 +20,9 @@ namespace xsimd
 
         template <class T, class Arch>
-        struct simd_register;
+        struct simd_register
```
I <3 all of your changes. Can you explain that one?

tomjnixon replied:
Ha, yes. Without that, the `static_assert` in `batch` never fires, because `simd_register` is not defined for unsupported architectures (there's only a forward declaration, which can't be inherited from).

That's why i'm not sure about it, as it means that `simd_register` can be constructed for unsupported architectures. I couldn't think of any way around this except for making it's constructor private, but that feels messy.

serge-sans-paille replied:
What about putting the same static_assert in `simd_register` then?

tomjnixon replied:
Oh yeah, that works. I tried this originally but got confused.

The difference now from the original is if you try to construct a batch with a bad type; you used to get

```
invalid use of incomplete type ‘struct xsimd::types::simd_register<double*, xsimd::sse2>’
```

whereas now you get

```
no type named ‘register_type’ in ‘struct xsimd::types::simd_register<double*, xsimd::sse2>’
```
I think it's fine; `simd_register` is presumably an internal thing anyway.

tomjnixon commented:
In this batch of changes I've:

- improved avx512 detection (tested with intel SDE)
- updated commit message in "don't check for supported architectures in dispatch" to mention, not close issue #633, as that might need more work
- moved the architecture check from `batch` to `simd_register`

I've also implemented the per-architecture build thing mentioned in #633 in my application using this -- seems to work ok!

The only thing missing is a way to get the maximum required alignment from an `arch_list`; I'll put in a separate PR for that.

tomjnixon commented:
Cheers!
